libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/introspection/introspection_reader.kt

```diff
@@ -0,0 +1,556 @@
+    )
+  }
+
+  private fun findInterfacesImplementedBy(name: String): List<String> {
+    return introspectionSchema.__schema
+        .types
+        .filter {
+          it.kind == RTypeKind.INTERFACE
+        }
+        .filter {
+          it.possibleTypes?.map { it.name }?.contains(name) == true
+        }
+        .map {
+          it.name
+        }
+  }
+
+  private fun RTypeFull.toGQLEnumTypeDefinition(): GQLEnumTypeDefinition {
+    return GQLEnumTypeDefinition(
+        description = description.unwrapDescription(name),
+        name = name,
+        enumValues = enumValues?.map { it.toGQLEnumValueDefinition() }
+            ?: throw ConversionException("Enum '$name' does not define any value"),
+        directives = emptyList()
+    )
+  }
+
+  private fun REnumValue.toGQLEnumValueDefinition(): GQLEnumValueDefinition {
+    return GQLEnumValueDefinition(
+        description = description.unwrapDescription(name),
+        name = name,
+        directives = makeDirectives(deprecationReason.unwrapDeprecationReason(name))
+    )
+  }
+
+  private fun Optional<String?>.unwrapDeprecationReason(name: String): String? {
+    return if (this is Optional.Absent) {
+      println("Apollo: $name is missing 'deprecationReason', double check your introspection query")
+      null
+    } else {
+      getOrThrow()
+    }
+  }
+
+  private fun RTypeFull.toGQLInterfaceTypeDefinition(): GQLInterfaceTypeDefinition {
+    return GQLInterfaceTypeDefinition(
+        name = name,
+        description = description.unwrapDescription(name),
+        fields = fields?.map { it.toGQLFieldDefinition() } ?: throw ConversionException("interface '$name' did not define any field"),
+        implementsInterfaces = interfaces?.mapNotNull { it.name } ?: emptyList(),
+        directives = emptyList()
+    )
+  }
+
+  private fun RField.toGQLFieldDefinition(): GQLFieldDefinition {
+    val args = if (args is Optional.Absent) {
+        println("Apollo: $name.args is missing, double check your introspection query")
+        emptyList()
+    } else {
+      args.getOrThrow()
+    }
+
+    return GQLFieldDefinition(
+        name = name,
+        description = description.unwrapDescription(name),
+        arguments = args.map { it.toGQLInputValueDefinition() },
+        directives = makeDirectives(deprecationReason.unwrapDeprecationReason(name)),
+        type = type.toGQLType()
+    )
+  }
+
+  private fun String?.toGQLValue(): GQLValue? {
+    if (this == null) {
+      // no default value
+      return null
+    }
+    try {
+      return Buffer().writeUtf8(this).parseAsGQLValue().getOrThrow()
+    } catch (e: Exception) {
+      throw ConversionException("cannot convert $this to a GQLValue")
+    }
+  }
+
+  /**
+   * This assumes the case where `deprecationReason == null && isDeprecated` is not valid
+   * which is unclear from reading the spec as the below seems allowed:
+   *
+   * ```
+   * type Query {
+   *   foo: Int @deprecated(reason: null)
+   * }
+   * ```
```

Comment:
Could we handle this case by passing `isDeprecated` + `deprecationReason`?

Reply:
The whole IR is based on the premise that `deprecationReason != null` implies `isDeprecated = true`. Kotlin `@Deprecated` annotation also requires a message. I'm still unclear what the use case is for `@deprecated(reason: null)`. I'll add a comment there